nixos/modules/services/web-apps/shiny-server.nix

```diff
@@ -0,0 +1,129 @@
+
+  generatedConfig = pkgs.writeText "shiny-server.conf" ''
+    run_as ${cfg.user};
+    server {
+      listen ${toString cfg.port};
+      location / {
+        site_dir ${cfg.siteDir};
+        log_dir ${cfg.logDir};
+        bookmark_state_dir ${cfg.stateDir};
+        directory_index ${boolToString cfg.directoryIndex};
+      }
+    }
+    ${cfg.extraConfig}
+  '';
+in
+{
+  options = {
+    services.shiny-server = {
+      enable = mkEnableOption "Enable server to host R Shiny applications.";
+
+      package = mkOption {
+        default = pkgs.shiny-server.override {
+          extraRPackages = cfg.extraPackages;
+        };
+        defaultText = ''pkgs.shiny-server.override {
+          extraRPackages = cfg.extraPackages;
+        };
+        '';
+        description = "Shiny Server package";
+      };
+
+      extraPackages = mkOption {
+        default = [];
+        example = ''with pkgs.rPackages; [ ggplot2 data_table ];'';
+        description = ''
+          Extra packages to include in the rWrapper used by shiny-server.
+          Use this to enable additional R packages.
+        '';
+      };
+
+      port = mkOption {
+        type = types.int;
+        default = 3838;
+        description = ''
+          Port number Shiny Server will be listening to.
+        '';
+      };
+
+      directoryIndex = mkOption {
+        type = types.bool;
+        default = true;
+        description = "Whether directory content is displayed";
+      };
+
+      user = mkOption {
+        default = "shiny";
+        description = "User to run Shiny Server as";
+      };
+
+      siteDir = mkOption {
+        type = types.path;
+        default = "${cfg.package}/shiny-server/samples";
+        defaultText = ''''${cfg.package}/shiny-server/samples'';
+        description = "Directory containing shiny apps to serve";
+      };
+      logDir = mkOption {
+        type = types.path;
+        default = "/var/lib/shiny/log";
+        description = "Directory to write server logs";
+      };
+      stateDir = mkOption {
+        type = types.path;
+        default = "/var/lib/shiny/state";
+        description = "Directory to store server state (bookmarks)";
+      };
+
+      extraConfig = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Extra lines to be added to the generated configuration file.
+        '';
+      };
+
+      configFile = mkOption {
+        type = types.path;
+        default = generatedConfig;
+        defaultText = "generatedConfig";
+        description = ''
+          Overridable configuration file. Defaults is the one generated.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    users.extraUsers = optionalAttrs (cfg.user == "shiny") (singleton {
+      name = cfg.user;
+      uid = config.ids.uids.shiny;
+      group = cfg.user;
+      description = "Shiny Server user";
+    });
+    users.extraGroups = optionalAttrs (cfg.user == "shiny") (singleton {
```

Comment:
extraUsers is now users, and extraGroups is just groups now. Also isn't this doing:

```
users.users = [ { name = "shiny" ... } ];
```

?

I'm surprised that works, as we should end up with: `users.users.shiny = { ... }`

Reply:
Interesting. Changed it and it looks like what you expect now:

nix-repl> :a (import ./nixos/default.nix { configuration = { config.services.shiny-server.enable = true; }; })
Added 5 variables.

nix-repl> config.users.users.shiny 
{ _module = { ... }; createHome = false; cryptHomeLuks = null; description = "Shiny Server user"; extraGroups = [ ... ]; group = "shiny"; hashedPassword = null; home = "/var/empty"; initialHashedPassword = null; initialPassword = null; isNormalUser = false; isSystemUser = false; name = "shiny"; openssh = { ... }; packages = [ ... ]; password = null; passwordFile = null; shell = «derivation /nix/store/v5qiscz1ypclkak40awlr50a01bzwdjm-shadow-4.5.drv»; subGidRanges = [ ... ]; subUidRanges = [ ... ]; uid = 285; useDefaultShell = false; }